### PR TITLE
[SPARK-29046][SQL] Fix NPE in SQLConf.get when active SparkContext is stopping

### PR DESCRIPTION
**Pull request by HeartSaVioR**
# What changes were proposed in this pull request?

This patch fixes the bug regarding NPE in SQLConf.get, which is only possible when SparkContext._dagScheduler is null due to stopping SparkContext. The logic doesn't seem to consider active SparkContext could be in progress of stopping.

Note that it can't be encountered easily as `SparkContext.stop()` blocks the main thread, but there're many cases which SQLConf.get is accessed concurrently while SparkContext.stop() is executing - users run another threads, or listener is accessing SQLConf.get after dagScheduler is set to null (this is the case what I encountered.)

### Why are the changes needed?

The bug brings NPE.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added new UT to verify NPE doesn't occur. Without patch, the test fails with throwing NPE.

### CODE REVIEW

**Review by srowen — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (vanzin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: nit: we use `{ sc => ... }` for closures. Also later.

**Comment by HeartSaVioR** [comment]
retest this, please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: Hi, All.

Ur, I'm wondering if this is safe in concurrent testing environments?
This might be the root cause of the current outage in Apache Spark Jenkins.
After this PR, all Jenkins on `master` branch never succeeds.
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/

Comment (vanzin) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: Given this is late Friday, just revert this if you thinks it's breaking things.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: Yes. I've been investigating this and the other commits. But, this is the usual suspect given the error. Since there has been broken over 2 days and it seems that we cannot fix this during weekend, I'll revert this.

**Comment by dongjoon-hyun** [comment]
Hi, All.
Since Jenkins will be reset at midnight, I'll revert this after Jenkins reset around Tomorrow 00:15AM.

**Comment by dongjoon-hyun** [comment]
FYI, this is reverted via https://github.com/apache/spark/commit/13b77e52d2315824b63758887b12b84360aa911e .

**Comment by dongjoon-hyun** [comment]
After reverting, Jenkins jobs are recovered. Please make another PR for this.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: Thanks for finding. I haven't noticed concurrent tests are executed, as there're many tests leveraging single object like SparkContext. I'll just exclude test code in new PR as I don't see other way to test this.

Btw, do we run different options between PR and master build? I'm curious about the reason - as we tend to guess PR build pass as "OK".

Comment (srowen) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: I think the issue is probably Maven vs SBT testing differences. I personally would prefer to standardize on the Maven build for everything.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: @HeartSaVioR  . PRBuilder is the same. PRBuilder has been very flaky last two days. I hit the same failure on my independent PRs too many times. That was the reason I investigated this and reached here to ask help.

@srowen . According to the Jenkins status, All Jenkins job combinations failed. This is irrelevant to SBT and Maven difference.

- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/

**Comment by dongjoon-hyun** [comment]
To committers.
Since 3.0.0-preview is coming, please take a look the Jenkins status after your merging.
We can notice this easily because this causes consecutive outages across over all Jenkins combinations from Sep 11, 2019 9:50PM to Sep 14, 2019 12:16AM without fixing the root causes. Also, many irrelevant PRs suffered during that period due to the flakiness.

**Comment by HyukjinKwon** [comment]
@dongjoon-hyun, I think everybody knows and does it. It's been public holidays in KST and I guess it's weekends - so the action was a  it delayed. Commit / revert thing can happen and I don't think there's particularly bad pattern here.

**Comment by HyukjinKwon** [comment]
> @dongjoon-hyun, I think everybody knows and does it. It's been public holidays in KST and I guess it's weekends - so the action was a bit delayed. Commit / revert thing can happen and I don't think there's particularly bad pattern here.

**Comment by dongjoon-hyun** [comment]
Of course, I knew that it was due to a big holiday for Lunar calendar. 😄

**Comment by HyukjinKwon** [comment]
Thanks all for taking care of this.

**Pull request closed without merging**
